image-registry/automate-pruning.md

```diff
@@ -0,0 +1,325 @@
+
+#### Pruning of in-use images
+
+Risk: Pruner is not aware of a particular image, resulting in it being pruned
+prior to use.
+
+Mitigation: None at present - this is an inherent problem within the current
+pruning process.
+
+## Design Details
+
+### API
+
+A separate CRD will be created with the following `spec` and `status` fields:
+
+```yaml
+spec:
+  schedule: "*/1 * * * *"
+  suspend: false
+  keepTagRevisions: 3
+  keepYoungerThan: 60m
+  resources: {}
+  affinity: {}
+  nodeSelector: {}
+  tolerations: {}
+  startingDeadlineSeconds: 60
+  history:
+    successfulJobsHistoryLimit: 3
+    failedJobsHistoryLimit: 1
+status:
+  observedGeneration: 2
+  conditions:
+  - type: Available
+    status: "True"
+    lastTransitionTime: 2019-10-09T03:13:45
+    reason: Ready
+    message: "Periodic image pruner has been created."
+  - type: Scheduled
+    status: "True"
+    lastTransitionTime: 2019-10-09T03:13:45
+    reason: Scheduled
+    message: "Image pruner job has been scheduled."
+  - type: Failed
+    staus: "False"
+    lastTransitionTime: 2019-10-09T03:13:45
+    reason: Succeeded
+    message: "Most recent image pruning job succeeded."
+```
+
+#### Spec
+
+1. `schedule:` Cron formatted schedule, defaults to daily at midnight for new
+clusters. Required field
+2. `suspend:` if `true`, CronJob running pruning is suspended. Optional,
+   default false.
+3. `keepTagRevisions:` number of revisions per tag to keep. Optional, default 3
+   if not set.
+4. `keepYoungerThan:` retain images younger than this duration. Optional,
+   default 60m if not set.
+5. `resources:` standard Pod resource requests and limits. Optional.
+6. `affinity:` standard Pod affinity. Optional.
+7. `nodeSelector:` standard Pod node selector. Optional
+8. `tolerations:` standard Pod tolerations. Optional
+9. `startingDeadlineSeconds:` start deadline for `CronJob`. Optional.
+10. `history:` sets the job history to be retained on the cluster. Object is
+    optional.
+    1. `successfulJobsHistoryLimit:` maximum number of successful jobs to
+       retain. Must be >= 1 to ensure metrics are reported.
+    2. `failedJobsHistoryLimit:` maximum number of failed jobs to retain. Must
+       be >= 1 to ensure metrics are reported.
+
+#### Status
+
+1. `observedGeneration:`: generation observed by the operator
+2. `conditions:` standard condition objects with the following types:
+   1. `Available` - indicates if the pruning job has been created.
+   Reasons can be `Ready` or `Error`.
+   2. `Scheduled` - indicates if the next pruning job has been scheduled.
+   Reasons can be `Scheduled`, `Suspended`, or `Error`.
+   3. `Failed` - indicates if the most recent pruning job failed.
+
+### Fixed Configuration
+
+The following settings for the pruning `CronJob` cannot be configured:
+
+1. `concurrencyPolicy: Forbid` - this prevents concurrent cron jobs from
+running.
+2. `suspend` - based on the provided `schedule`. If empty `suspend = true`.
```

Comment:
this seems to conflict with the api described above which says the schedule must be set, and suspend is a valid field to configure.